dropshot/src/lib.rs

```diff
@@ -275,6 +270,178 @@
+ * all the amphibians in sorted order, then all the mammals, then all the
+ * reptiles.  The main requirement is that the combination of fields used for
+ * pagination must be unique.  We cannot paginate by the animal's class alone.
+ * (To see why: there are over 6,000 mammals.  If the page size is, say, 1000,
+ * then the page_token would say `"mammal"`, but there's not enough information
+ * there to see where we are within the list of mammals.  It doesn't matter
+ * whether there are 2 mammals or 6,000 because clients can limit the page size
+ * to just one item if they want and that ought to work.)
+ *
+ *
+ * ### Dropshot interfaces for pagination
+ *
+ * We can think of pagination in two parts: the input (handling the pagination
+ * query parameters) and the output (emitting a page of results, including the
+ * page token).
+ *
+ * For input, a paginated API endpoint's handler function should accept a
+ * [`Query`]`<`[`PaginationParams`]`<ScanParams, PageSelector>>`, where
+ * `ScanParams` is a consumer-defined type specifying the parameters of the scan
+ * (typically including the sort fields, sort order, and filter options) and
+ * `PageSelector` is a consumer-defined type describing the page token.  The
+ * PageSelector will be serialized to JSON and base64-encoded to construct the
+ * page token.  This will be automatically parsed on the way back in.
+ *
+ * For output, a paginated API endpoint's handler function can return
+ * `Result<`[`HttpResponseOkObject`]<[`ResultsPage`]`<T>, HttpError>` where `T:
+ * Serialize` is the item listed by the endpoint.  You can also use your own
+ * structure that contains a [`ResultsPage`] (possibly using
+ * `#[serde(flatten)]`), if that's the behavior you want.
+ *
+ * There are several complete, documented examples in the "examples" directory.
+ *
+ *
+ * ### Advanced usage notes
+ *
+ * It's possible to accept additional query parameters besides the pagination
+ * parameters by having your API endpoint handler function take two different
+ * arguments using `Query`, like this:
+ *
+ * ```
+ * use dropshot::ExtractedParameter;
+ * use dropshot::HttpError;
+ * use dropshot::HttpResponseOkObject;
+ * use dropshot::PaginationParams;
+ * use dropshot::Query;
+ * use dropshot::RequestContext;
+ * use dropshot::ResultsPage;
+ * use dropshot::endpoint;
+ * use serde::Deserialize;
+ * use std::sync::Arc;
+ * # use serde::Serialize;
+ * # #[derive(Debug, Deserialize, ExtractedParameter)]
+ * # enum MyScanParams { A };
+ * # #[derive(Debug, Deserialize, ExtractedParameter, Serialize)]
+ * # enum MyPageSelector { A(String) };
+ * #[derive(Deserialize, ExtractedParameter)]
+ * struct MyExtraQueryParams {
+ *     do_extra_stuff: bool,
+ * }
+ *
+ * #[endpoint {
+ *     method = GET,
+ *     path = "/list_stuff"
+ * }]
+ * async fn my_list_api(
+ *     rqctx: Arc<RequestContext>,
+ *     pag_params: Query<PaginationParams<MyScanParams, MyPageSelector>>,
+ *     extra_params: Query<MyExtraQueryParams>,
```

Comment:
I can definitely see how it's a little strange.  On the other hand, it would seem a little strange to me if it _didn't_ work -- that would imply that extractors somehow mutate the request or are otherwise intertwined instead of being self-contained things that construct a higher-level view of something in the request.

With the serde bug, I don't know another way to support taking non-pagination query parameters, which seems pretty useful.

I thought about whether we can avoid parsing the querystring twice.  Unfortunately, we don't know until we've parsed the querystring whether it's the `WhichPage::First` case, in which case _all_ of the query parameters should be deserialized into the user's struct.  But once we've parsed it to determine we're in the `WhichPage::First` case, we don't have a way to take the representation we've got and parse it into the user's type.  The only way we have to get to the user's struct is to parse the querystring again.  Alternatively, we could define an intermediate representation (e.g., `BTreeMap<String, String>`) and define a serde Deserializer for that representation that can reconstitute the user's type, but this doesn't feel worth the work right now just to avoid a second parse.